app/assets/javascripts/tsheets_integration.js

```diff
@@ -0,0 +1,65 @@
+var tsheets = (function($) {
```

Comment:
Sam, I should say that i like how you write JS - neatly organized :-)

Reply:
The suggestion for react was more because Connor was familiar with it (plus it's very easy to find new devs who are familiar with it). I'm not specifically a fan of react, I like it but I do think there are better alternatives out there, it's just that it's so easy to find developers so is a good practical solution in my eyes.

React is not that big itself actually, nothing like Ember or Angular, unless you want to use it for an SPA (which we wouldn't I think) and add in things such as Redux and React Router.

Riot.js looks really nice though, pretty much a smaller version of react. 

The reason to switch to something like react would be simply because a large JS codebase easily becomes a mess as you say - componentising everything helps to rationalise the code (although you still need to be careful about component structure to avoid _that_ getting into a mess).

So in summary, react would be a practical solution in my opinion, but I'm open to other suggestions if everyone is in agreement.

Reply:
I'm not sure how React would help to have a sane codebase, Sam. Can you share a bit more on this? I would pick something like TypeScript to help with huge JS code base. I only recently picked it up (couple of months), but so far very happy with it.

Reply:
Sorry I forgot to reply to this one. React (and Riot.js) help in that they isolate the code. Basically the js for a particular component is tied to the component and will only get run if that component is in play.

Compare this to normal js in a rails app where you have global variable hell, hundreds of `$(document).ready` calls (for every possible page) all being run on every single page. You need to make sure your selectors don't collide, and be careful about executing code on the wrong page or in the wrong context. It's like Rails executing all controller actions (for every controller) on each request, and you have to check in each action whether the code should be running.

React, Riot, Vue (I believe, I've not used Vue) all resolve this in the same way by tying code to the component.

Although as mentioned, you still need to give thought to component structure to keep the code base sane.

Personally I like the idea of Elm, but I don't think it's wise to introduce something that very few people know. ;-)

I've not used TypeScript though, but it's only a language isn't it, or does it offer some sort of framework for organising code?

Reply:
@cheeseweasel I've picked up typescript, because it's a very small subset of feature on top of ES6 JS that make it more easy to mantain (at least for me). This post goes into details why Angular 2.0 was written in TypeScript -- https://vsavkin.com/writing-angular-2-in-typescript-1fa77c78d8e8